### PR TITLE
Simplify version detection 

### LINKED ISSUE #324: Simplify version detection
The current code fetches version first using the root path and then using `/:workspace/kong` if the first one fails. This needs to be refactored and simplified:
- if deck is run with empty workspace, read version from Root()
- if deck is run with a non-empty workspace, read version from /<workspace-name>/kong

### PR DESCRIPTION
**Pull request by mmorel-35**
Fixes #324

### CODE REVIEW

**Comment by mmorel-35** [comment]
The access to the version of kong shall be moved to the go-kong but there is a need for the client to be aware of the use of a workspace. There could also be a service to expose every versioned capabilities of kong like tags for credentials in deck or mtls authentication in kubernetes ingress controller. This way the semver dependency would only be required in go-kong

**Review by mflendrich — COMMENTED** [review]
Thanks @mmorel-35 for looking into this.

> There could also be a service to expose every versioned capabilities of kong

I think that would be the best approach for future maintainability; @rainest @shaneutt what do you think?

File: cmd/common.go [hunk 1]
```diff
@@ -177,33 +176,27 @@ func kongVersion(config utils.KongClientConfig) (semver.Version, error) {
...
-			nil)
+
+	if len(config.Workspace) > 0 {
+		req, err := client.NewRequest("GET", "/kong", nil, nil)
 		if err != nil {
 			return semver.Version{}, err
 		}
 		var resp map[string]interface{}
-		_, err = client.Do(nil, req, &resp)
+		_, err = client.Do(context.TODO(), req, &resp)
```
[the remark this answers is not in the document]
Comment (mmorel-35) [reply cmd/common.go]: Done, I also did it fo other functions cf.  9bfeb5d and 3e998d0

**Review by mflendrich — COMMENTED** [review]
See the comments/suggestions below, and please ensure that this works with Kong OSS and Enterprise (both with workspaces used with RBAC preventing access to the non-workspaced root and not). Unfortunately we don't have automated regression tests capturing that, but we need to ensure that we haven't broken anything in those scenarios before merging.

**Comment by hbagdi** [comment]
> Unfortunately we don't have automated regression tests capturing that, but we need to ensure that we haven't broken anything in those scenarios before merging.

@mflendrich Did you mean to imply that we should add that in here before merging in?
I think adding integration tests like these requires more thought and should probably be out of scope from this PR.

**Comment by mmorel-35** [comment]
Following my suggestion from last week I created a [PR](https://github.com/Kong/go-kong/pull/48) on go-kong. So that the `kongVersion()` is hold by the client. The the version is a string and has the format of the versions of the kic so it has more informations than what deck is actually showing. Also the test on the version to know if it's greater than 1.4.0 is in the response so deck don't need to do it and so don't need to have semver as a direct dependency.

**Comment by mflendrich** [comment]
@mmorel-35, with the https://github.com/Kong/go-kong/pull/65 PR moving the API machinery into go-kong, which parts of this PR are still relevant? Should we close it and follow up on https://github.com/Kong/go-kong/pull/65 by using that API when that PR gets merged in?

On integration testing: we're tracking the direction for integration testing in deck in https://github.com/Kong/deck/issues/63. I'll post this PR there as prior work, but I don't think we're ready to merge the test yet.

That being said, I think the best way forward would be to close this PR and follow up with:
- (when the go-kong PR with the information service gets merged and released) using the go-kong information service for version retrieval
- (when we make up our mind about integration testing) implement the integration test using this PR as reference

**Comment by mmorel-35** [comment]
I agree!

**Pull request closed without merging**